File: lib/models/user_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class UserModel {
  // Fields stored in Firebase Authentication and Firestore
  final String uid;
  final String email;
  final String name;
  final String role; // 'Donor', 'Volunteer', or 'Admin'
  final String
  status; // e.g., 'active', 'pending_verification' (for volunteers)
  final Timestamp createdAt;

  // Donor-specific fields (optional)
  final String? organizationId;
  final String? primaryLocation;

  // Volunteer-specific fields (optional)
  final int? tasksCompleted;
  final double? rating;

  UserModel({
    required this.uid,
    required this.email,
    required this.name,
    required this.role,
    required this.status,
    required this.createdAt,
    this.organizationId,
    this.primaryLocation,
    this.tasksCompleted,
    this.rating,
  });

  // Factory constructor to create a UserModel from a Firestore document
  factory UserModel.fromMap(Map<String, dynamic> data) {
    return UserModel(
      uid: data['uid'] as String,
      email: data['email'] as String,
      name: data['name'] as String,
      role: data['role'] as String,
      status: data['status'] as String? ?? 'active',
      createdAt: data['createdAt'] as Timestamp? ?? Timestamp.now(),
      organizationId: data['organizationId'] as String?,
      primaryLocation: data['primaryLocation'] as String?,
      tasksCompleted: data['tasksCompleted'] as int?,
      rating: data['rating'] is int
          ? (data['rating'] as int).toDouble()
          : data['rating'] as double?,
    );
  }

  // Method to convert the UserModel to a map for writing to Firestore
  Map<String, dynamic> toMap() {
    return {
      'uid': uid,
      'email': email,
      'name': name,
      'role': role,
      'status': status,
      'createdAt': createdAt,
      'organizationId': organizationId,
      'primaryLocation': primaryLocation,
      'tasksCompleted': tasksCompleted,
      'rating': rating,
    };
  }
}

File: server/models/userModel.js
// Import the Mongoose library 
const mongoose = require("mongoose");
// Extract the Schema constructor from Mongoose
const { Schema } = mongoose;

// Define the structure for a User document in the database
const UserSchema = new Schema({
  // Firebase UID of the user (comes from Firebase Auth)
  uid: { type: String, required: true, unique: true },

  // Email address of the user (must be unique across the system)
  email: { type: String, required: true, unique: true },

  // Full name of the user
  name: { type: String, required: true },

  // Defines whether user is a Donor, Volunteer, or Admin
  // ENUM restricts values so invalid roles can't be stored
  role: { type: String, required: true, enum: ["Donor", "Volunteer", "Admin"] },

  // Active/inactive user status (used by Admin to deactivate accounts)
  status: { type: String, default: "active" },

  // Automatically stores when the user record was created
  createdAt: { type: Date, default: Date.now },

  // Stores last update time for the user record
  updatedAt: { type: Date, default: Date.now },
});

// Compound index for efficient queries when filtering users by role + status
UserSchema.index({ role: 1, status: 1 });

// Convert the schema into a Mongoose model to access CRUD functions
const User = mongoose.model("User", UserSchema);

// Creates a new user in the database
const createUser = async (userData) => {
  const newUser = new User(userData);
  return await newUser.save();
};

// Fetch a user using their Firebase UID
const getUserByUid = async (uid) => {
  return await User.findOne({ uid });
};

// Fetch a user using their email address
const getUserByEmail = async (email) => {
  return await User.findOne({ email });
};

// Get users based on role (e.g., all Volunteers) but only active ones
const getUsersByRole = async (role) => {
  return await User.find({ role, status: "active" });
};

// Update a user’s data by UID and automatically refresh the updatedAt field
const updateUser = async (uid, updateData) => {
  return await User.findOneAndUpdate(
    { uid },
    { ...updateData, updatedAt: Date.now() },
    { new: true } // returns updated document instead of old one
  );
};

// Export model and helper functions for use in controllers or services
module.exports = {
  User,
  createUser,
  getUserByUid,
  getUserByEmail,
  getUsersByRole,
  updateUser,
};
